### ISSUE TITLE
[native-image] Unexpected number of interfaces for annotation proxy class

### ISSUE THREAD

**Issue by gwenneg** [oracle/graal#1941]
Hi!

I'm getting the following error while generating a GraalVM `19.3.0` native image using Quarkus `master` and JDK 11:

```
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] /Library/Java/JavaVirtualMachines/graalvm-ce-java11-19.3.0/Contents/Home/bin/native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -jar quarkus-integration-test-hibernate-validator-999-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -J-Xrunjdwp:transport=dt_socket,address=5005,server=y,suspend=y -H:+PrintAnalysisCallTree -Dsun.misc.ProxyGenerator.saveGeneratedFiles=true -H:+AddAllCharsets -H:EnableURLProtocols=http -H:+JNI --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quarkus-integration-test-hibernate-validator-999-SNAPSHOT-runner
Listening for transport dt_socket at address: 5005
[quarkus-integration-test-hibernate-validator-999-SNAPSHOT-runner:10519]    classlist:  17,089.81 ms
[quarkus-integration-test-hibernate-validator-999-SNAPSHOT-runner:10519]        (cap):   6,451.99 ms
[quarkus-integration-test-hibernate-validator-999-SNAPSHOT-runner:10519]        setup:  11,578.56 ms
20:59:31,393 INFO  [org.hib.val.int.uti.Version] HV000001: Hibernate Validator 6.1.0.Final
20:59:31,701 WARN  [org.hib.val.int.pro.jav.JavaBeanExecutable] HV000254: Missing parameter metadata for MyEnumWithValidatorAnnotation(String, int), which declares implicit or synthetic parameters. Automatic resolution of generic type information for method parameters may yield incorrect results if multiple parameters have the same erasure. To solve this, compile your code with the '-parameters' flag.
20:59:36,315 INFO  [org.jbo.threads] JBoss Threads version 3.0.0.Final
[quarkus-integration-test-hibernate-validator-999-SNAPSHOT-runner:10519]     analysis:  45,828.00 ms
Fatal error: com.oracle.svm.core.util.VMError$HostedError: Unexpected number of interfaces for annotation proxy class.
	at com.oracle.svm.core.util.VMError.guarantee(VMError.java:85)
	at com.oracle.svm.hosted.annotation.AnnotationSupport.findAnnotationInterfaceTypeForMarkedAnnotationType(AnnotationSupport.java:543)
	at com.oracle.svm.hosted.annotation.AnnotationSubstitutionType.<init>(AnnotationSubstitutionType.java:46)
	at com.oracle.svm.hosted.annotation.AnnotationSupport.getSubstitution(AnnotationSupport.java:199)
	at com.oracle.svm.hosted.annotation.AnnotationSupport.lookup(AnnotationSupport.java:156)
	at com.oracle.graal.pointsto.infrastructure.SubstitutionProcessor$ChainedSubstitutionProcessor.lookup(SubstitutionProcessor.java:113)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.lookupAllowUnresolved(AnalysisUniverse.java:202)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.lookup(AnalysisUniverse.java:182)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.lookup(AnalysisUniverse.java:75)
	at com.oracle.graal.pointsto.infrastructure.UniverseMetaAccess$1.apply(UniverseMetaAccess.java:52)
	at com.oracle.graal.pointsto.infrastructure.UniverseMetaAccess$1.apply(UniverseMetaAccess.java:49)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1737)
	at com.oracle.graal.pointsto.infrastructure.UniverseMetaAccess.lookupJavaType(UniverseMetaAccess.java:84)
	at com.oracle.graal.pointsto.meta.AnalysisMetaAccess.lookupJavaType(AnalysisMetaAccess.java:47)
	at com.oracle.svm.hosted.ameta.HostedDynamicHubFeature.replace(HostedDynamicHubFeature.java:52)
	at com.oracle.graal.pointsto.meta.AnalysisUniverse.replaceObject(AnalysisUniverse.java:534)
	at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.replaceObject(AnalysisConstantReflectionProvider.java:201)
	at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.interceptValue(AnalysisConstantReflectionProvider.java:172)
	at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.readValue(AnalysisConstantReflectionProvider.java:101)
	at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.readFieldValue(AnalysisConstantReflectionProvider.java:77)
	at com.oracle.graal.pointsto.ObjectScanner.scanField(ObjectScanner.java:175)
	at com.oracle.graal.pointsto.ObjectScanner.doScan(ObjectScanner.java:344)
	at com.oracle.graal.pointsto.ObjectScanner.access$400(ObjectScanner.java:62)
	at com.oracle.graal.pointsto.ObjectScanner$3$1.run(ObjectScanner.java:414)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:171)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

When I placed a breakpoint in debug at `AnnotationSupport.java:543`, I could see that there's only one interface which is `HotSpotType<Ljava/lang/Override;, resolved>` for the failing field but I couldn't find which annotation is involved so far, except that it is proxied: `Lcom/sun/proxy/$Proxy152;`.

Would you have any idea that could help me identify the cause of this error?

**Comment by jaikiran** [comment]
@gwenneg, Having looked at the `AnnotationSupport` code in substratevm, I think this is more of a bug in Quarkus and Gizmo combined. What seems to be happening is that we (Quarkus + Gizmo) seem to be generating bytecode for a class which ends up registering `java.lang.Override` as an annotation on the generated class. The `java.lang.Override` is of type `@Retention(RetentionPolicy.SOURCE)` and as per the contract of `RetentionPolicy.SOURCE`:

```
/**
     * Annotations are to be discarded by the compiler.
     */
    SOURCE,
```
The annotation with that retention type should never make it into the bytecode. 

The `AnnotationSupport` in substratevm specifically/intentionally relies on this expectation[1] and our invalid (IMO) bytecode is causing it's logic here[2] to incorrectly consider this `Proxy` (which only has (the incorrect) `Override` annotation applied on it) to be considered for annotation substitution.

[1] https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/annotation/AnnotationSupport.java#L99
[2] https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/annotation/AnnotationSupport.java#L615

**Comment by jaikiran** [comment]
The only mystery was, why was this only happening on Java 11 and how the `@Override` on a method ends up creating a `Proxy` class. That's now solved by looking at the `sun.reflect.annotation.AnnotationParser` code. This code is different in Java 8 and 11 and in Java 11 it ends up creating a `Proxy` while parsing the annotations https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/sun/reflect/annotation/AnnotationParser.java#L301.

So in short, I think this actually is something that we should fix in Quarkus and that part you have already taken care of in https://github.com/quarkusio/quarkus/pull/5999

**Comment by gwenneg** [comment]
Thanks for the `sun.reflect.annotation.AnnotationParser` analysis @jaikiran.

I only have one question for the GraalVM team about this issue. Even if it was indeed an issue resulting from a wrong bytecode generation in Quarkus, it took me a few hours to debug the GraalVM native image generation and trace the issue back to Quarkus. It would be great if this problem could be detected by `native-image` (from what I saw in debug, it's possible) and if an error pointing the user to the exact issue was thrown. What do you think about it @cstancu?

**Comment by cstancu** [comment]
@gwenneg while the error message can be improved that is just an internal assertion in SVM code and not something that should ever show up in user code. As @jaikiran correctly analysed the problem was caused by invalid bytecode generated by the framework. Pull requests to improve the debugging experience are welcome.
